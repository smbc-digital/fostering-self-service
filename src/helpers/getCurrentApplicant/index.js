import { Applicant, SECOND_APPLICANT_ROUTE } from 'constants'

const getCurrentApplicant = match => {
    if (match.params && match.params[0] === SECOND_APPLICANT_ROUTE) {
        return Applicant.SecondApplicant
    }

    return Applicant.FirstApplicant
}

export default getCurrentApplicant