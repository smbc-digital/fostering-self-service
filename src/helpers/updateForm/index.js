import { fetchWithTimeout } from '../index'

const TaskStatus = {
    None: 0,
    Completed: 1,
    NotCompleted: 2,
    CantStart: 3
}

const StageName = {
    HomeVisit: 0,
    Application: 1
}

const HomeVisitFormName = {
    TellUsAboutYourself: 0,
    YourEmploymentDetails: 1,
    LanguagesSpokenInYourHome: 2,
    YourPartnership: 3,
    YourFosteringHistory: 4,
    YourHealth: 5,
    TellUsAboutYourInterestInFostering: 6,
    YourHousehold: 7,
    ChildrenLivingAwayFromYourHome: 8
}

const ApplicationFormName = {
    References: 0,
    GpDetails: 1,
    AddressHistory: 2
}

const getHomeVisitUpdateEndpoint = form => {
    switch (form) {
        case HomeVisitFormName.TellUsAboutYourself:
            return '/fostering/home-visit/about-yourself'
        case HomeVisitFormName.YourEmploymentDetails:
            return '/fostering/home-visit/your-employment-details'
        case HomeVisitFormName.LanguagesSpokenInYourHome:
            return '/fostering/home-visit/languages-spoken-in-your-home'
        case HomeVisitFormName.YourFosteringHistory:
            return '/fostering/home-visit/your-fostering-history'
        case HomeVisitFormName.YourPartnership:
            return '/fostering/home-visit/partnership-status'
        case HomeVisitFormName.TellUsAboutYourInterestInFostering:
            return '/fostering/home-visit/interest-in-fostering'
        case HomeVisitFormName.YourHealth:
            return '/fostering/home-visit/about-your-health'
        case HomeVisitFormName.YourHousehold:
            return '/fostering/home-visit/household'
        case HomeVisitFormName.ChildrenLivingAwayFromYourHome:
            return '/fostering/home-visit/children-living-away-from-home'
        default:
            throw new Error('No matching endpoint for given form.')
    }
}

const getApplicationUpdateEndpoint = form => {
    switch (form) {
        case ApplicationFormName.References:
            return '/fostering/application/references'
        case ApplicationFormName.GpDetails:
            return '/fostering/application/gp-details'
        default:
            throw new Error('No matching endpoint for given form.')
    }
}

const getFormStageEndpoint = stage => {
    switch (stage) {
        case StageName.HomeVisit: 
            return '/fostering/home-visit'
        case StageName.Application:
            return '/fostering/application'
        default:
            throw new Error('No matching endpoint for given stage.')
    }
}

const reduceProperties = object => Object.keys(object).reduce((acc, property) => {
    return {
        ...acc,
        [property]: object[property].value
    }
}, {})

const updateForm = async (endpoint, formData) => {
    const parsedFormData = parseFormData(formData)

    const response = await fetchWithTimeout(endpoint, {
        method: 'PATCH',
        credentials: 'include',
        body: JSON.stringify(parsedFormData),
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        }
    }, 30000)

    if (!response.ok) {
        throw Error(response.statusText)
    }

    return await response.json()
}

const updateFormStatus = ({ form, stage, currentStatus, setStatus }) => {
    if (currentStatus === TaskStatus.None) {
        fetchWithTimeout(`${getFormStageEndpoint(stage)}/status`,
            {
                method: 'PATCH',
                credentials: 'include',
                body: JSON.stringify({
                    status: TaskStatus.NotCompleted,
                    form
                }),
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                }
            }, 30000)

        setStatus(TaskStatus.NotCompleted)
    }
}

const parseFormData = ({ firstApplicant, secondApplicant, familyReference, firstPersonalReference, secondPersonalReference, ...formData }) => {
    let parsedObject = reduceProperties(formData)

    if (firstApplicant) {
        parsedObject.firstApplicant = reduceProperties(firstApplicant)
    }

    if (secondApplicant) {
        parsedObject.secondApplicant = reduceProperties(secondApplicant)
    }

    if (familyReference) {
        parsedObject.familyReference = reduceProperties(familyReference)
        parsedObject.familyReference.address = familyReference.address.value
    }

    if (firstPersonalReference) {
        parsedObject.firstPersonalReference = reduceProperties(firstPersonalReference)
        parsedObject.firstPersonalReference.address = firstPersonalReference.address.value
    }

    if (secondPersonalReference) {
        parsedObject.secondPersonalReference = reduceProperties(secondPersonalReference)
        parsedObject.secondPersonalReference.address = secondPersonalReference.address.value
    }

    return parsedObject
}

const updateApplicationForm = async (form, formData) => {
    const endpoint = getApplicationUpdateEndpoint(form)

    return await updateForm(endpoint, formData)
}

const updateHomeVisitForm = async (form, formData) => {
    const endpoint = getHomeVisitUpdateEndpoint(form)

    return await updateForm(endpoint, formData)
}

export {
    TaskStatus,
    StageName,
    HomeVisitFormName,
    ApplicationFormName,
    updateFormStatus,
    parseFormData,
    updateApplicationForm,
    updateHomeVisitForm
}