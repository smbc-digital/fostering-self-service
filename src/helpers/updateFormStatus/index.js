export const TaskStatus = {
    None: 0,
    Completed: 1,
    NotCompleted: 2,
    CantStart: 3
}

export const FormName =
{
    TellUsAboutYourself: 0,
    YourEmploymentDetails: 1,
    LanguageSpokenInYourHome: 2,
    YourPartnership: 3,
    YourFosteringHistory: 4,
    YourHealth: 5,
    TellUsAboutYourInterestInFostering: 6,
    YourHousehold: 7,
    ChildrenLivingAwayFromYourHome: 7
}

export default function (form, currentStatus, setStatus) {
    if (currentStatus === TaskStatus.None) {
        fetch('/front-end-endpoint-update-status',
            {
                method: 'PATCH',
                credentials: 'include',
                body: JSON.stringify({
                    status: TaskStatus.NotCompleted,
                    form
                })
            })

        setStatus(TaskStatus.NotCompleted)
    }
}