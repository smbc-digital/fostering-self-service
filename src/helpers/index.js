import fetchWithTimeout from './fetchWithTimeout'
import { getPageRoute } from './pagehelper'
import getCurrentApplicant from './getCurrentApplicant'
import { updateFormStatus, updateForm, FormName, TaskStatus } from './updateForm'

export {
    fetchWithTimeout,
    getPageRoute,
    getCurrentApplicant,
    updateFormStatus,
    updateForm,
    FormName,
    TaskStatus
}