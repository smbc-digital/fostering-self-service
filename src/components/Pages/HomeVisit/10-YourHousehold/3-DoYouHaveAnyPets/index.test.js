import { React, mount, useContextMock, renderer } from 'helpers/SetupTest'
import DoYouHaveAnyPets from './index'
import * as helpers from 'helpers'

describe('DoYouHaveAnyPets', () => {
    const mockOnChangeStatus = jest.fn()
    beforeEach(() => {
        useContextMock.mockReturnValue({
            onChange: jest.fn(),
            onChangeStatus: mockOnChangeStatus,
            doYouHaveAnyPets: {
                value: '',
                isValid: false
            },
            petsInformation: {
                value: '',
                isValid: false
            },
            otherPeopleInYourHousehold: {
                value: [],
                isValid: false
            },
            anyOtherPeopleInYourHousehold: {
                value: false,
                isValid: false
            },
            statuses: {
                yourHouseholdStatus: 0
            }
        })
    })

    it('Should call history.push on submit', async () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        helpers.updateHomeVisitForm = jest.fn().mockReturnValue(Promise.resolve(0))

        // Act
        const wrapper = mount(<DoYouHaveAnyPets history={history} />)
        await wrapper.find('form').simulate('submit')
        await Promise.resolve()

        // Assert
        expect(history.push).toHaveBeenCalledWith(helpers.getPageRoute(17))
        expect(mockOnChangeStatus).toHaveBeenCalledWith('yourHouseholdStatus', 0)
    })

    it('Should call history.push on submit when goBack button pressed', async () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        helpers.updateHomeVisitForm = jest.fn().mockReturnValue(Promise.resolve(0))

        // Act
        const wrapper = mount(<DoYouHaveAnyPets history={history} />)
        await wrapper.find('Button').at(1).simulate('click')
        await Promise.resolve()
        
        // Assert
        expect(history.push).toHaveBeenCalledWith(helpers.getPageRoute(1))
        expect(mockOnChangeStatus).toHaveBeenCalledWith('yourHouseholdStatus', 0)
    })

    it('Should go to error page', () => {
        // Arrange
        const history = {
            push: jest.fn()
        }
        helpers.updateHomeVisitForm = jest.fn().mockImplementation(() => { throw new Error()})

        // Act
        const wrapper = mount(<DoYouHaveAnyPets history={history} />)
        wrapper.find('Button').at(1).simulate('click')
        
        // Assert
        expect(history.push).toHaveBeenCalledWith('/error')
    })

    it('Should show textarea when user has pets', () => {
        // Arrange
        useContextMock.mockReturnValue({
            onChange: jest.fn(),
            onChangeStatus: mockOnChangeStatus,
            doYouHaveAnyPets: {
                value: true,
                isValid: false
            },
            petsInformation: {
                value: '',
                isValid: false
            },
            otherPeopleInYourHousehold: {
                value: [],
                isValid: false
            },
            anyOtherPeopleInYourHousehold: {
                value: false,
                isValid: false
            },
            statuses: {
                yourHouseholdStatus: 0
            }
        })

        // Act
        const wrapper = mount(<DoYouHaveAnyPets history={history} />)
        
        // Assert
        expect(wrapper.find('textarea').exists()).toBe(true)
    })

    describe('snapshot', () => {
        it('renders correctly', () => {
            const tree = renderer
                .create(<DoYouHaveAnyPets history={{}}/>)
                .toJSON()

            expect(tree).toMatchSnapshot()
        })
    })
})