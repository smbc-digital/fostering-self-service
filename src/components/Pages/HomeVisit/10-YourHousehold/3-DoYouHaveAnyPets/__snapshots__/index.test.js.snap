// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DoYouHaveAnyPets snapshot renders correctly 1`] = `
Array [
  <h1>
    Your fostering journey
  </h1>,
  <h2>
    Your household
  </h2>,
  <form
    onSubmit={[Function]}
  >
    <h3
      className="no-margin-bottom"
    >
      Do you have any pets?
    </h3>
    <fieldset>
      <div
        className="input-radio-container"
        role="radiogroup"
      >
        <input
          checked={false}
          disabled={undefined}
          id="do-you-have-any-pets-yes"
          name="doYouHaveAnyPets"
          onChange={[Function]}
          type="radio"
          value="true"
        />
        <label
          htmlFor="do-you-have-any-pets-yes"
        >
          Yes
        </label>
      </div>
      <div
        className="input-radio-container"
        role="radiogroup"
      >
        <input
          checked={false}
          disabled={undefined}
          id="do-you-have-any-pets-no"
          name="doYouHaveAnyPets"
          onChange={[Function]}
          type="radio"
          value="false"
        />
        <label
          htmlFor="do-you-have-any-pets-no"
        >
          No
        </label>
      </div>
    </fieldset>
    <button
      className="button-primary"
      disabled={false}
      id={undefined}
      onClick={undefined}
      type="submit"
    >
      Save and next step
    </button>
    <p>
      <a
        href="#"
        onClick={[Function]}
      >
        Previous
      </a>
    </p>
    <button
      className="button-inverted left-chevron"
      disabled={false}
      id={undefined}
      onClick={[Function]}
      type="submit"
    >
      Save and go back to fostering area
    </button>
  </form>,
]
`;
